File: frontend/src/api/access.ts
import { api } from "./client";

export type AccessRequestStatus = "pending" | "approved" | "rejected" | "withdrawn";
export type ResourceType = "file_share" | "application" | "database" | "api";

export interface Permission {
  id: string;
  code: string;
  description: string;
}

export interface Role {
  id: string;
  code: string;
  name: string;
  description: string;
  is_privileged: boolean;
  owner_user_id?: string;
}

export interface RoleDetail extends Role {
  permissions: Permission[];
}

export interface RolesListResponse {
  items: Role[];
  total: number;
  page: number;
  page_size: number;
}

export interface UserRole {
  id: string;
  user_id: string;
  role_id: string;
  role: Role;
  granted_at: string;
  granted_by?: string;
  expires_at?: string;
  request_id?: string;
}

export interface AccessRequest {
  id: string;
  user_id: string;
  role_id: string;
  role: Role;
  justification: string;
  status: AccessRequestStatus;
  created_at: string;
  decided_at?: string;
  decided_by?: string;
  decision_comment?: string;
}

export interface AccessRequestsListResponse {
  items: AccessRequest[];
  total: number;
  page: number;
  page_size: number;
}

export interface PositionMatrixRow {
  position_id: string;
  position_code: string;
  position_name: string;
  role_ids: string[];
  roles: Role[];
}

export const accessApi = {
  // Permissions
  listPermissions: () => api.get<Permission[]>("/access/permissions"),

  // Roles
  listRoles: (params?: { search?: string; page?: number; page_size?: number }) =>
    api.get<RolesListResponse>("/access/roles", { params }),

  getRole: (id: string) => api.get<RoleDetail>(`/access/roles/${id}`),

  createRole: (body: { code: string; name: string; description?: string; is_privileged?: boolean }) =>
    api.post<Role>("/access/roles", body),

  updateRole: (id: string, body: { name?: string; description?: string; is_privileged?: boolean }) =>
    api.patch<Role>(`/access/roles/${id}`, body),

  deleteRole: (id: string) => api.delete(`/access/roles/${id}`),

  addPermissionToRole: (roleId: string, permissionId: string) =>
    api.post(`/access/roles/${roleId}/permissions/${permissionId}`),

  removePermissionFromRole: (roleId: string, permissionId: string) =>
    api.delete(`/access/roles/${roleId}/permissions/${permissionId}`),

  // User roles
  getUserRoles: (userId: string) => api.get<UserRole[]>(`/access/users/${userId}/roles`),

  assignRole: (userId: string, body: { role_id: string; expires_at?: string }) =>
    api.post<UserRole>(`/access/users/${userId}/roles`, body),

  revokeRole: (userId: string, userRoleId: string) =>
    api.delete(`/access/users/${userId}/roles/${userRoleId}`),

  // Permission check
  checkPermission: (userId: string, permissionCode: string) =>
    api.post<{ user_id: string; permission_code: string; allowed: boolean }>("/access/check", {
      user_id: userId,
      permission_code: permissionCode,
    }),

  // Access requests
  listRequests: (params?: { user_id?: string; status?: string; page?: number; page_size?: number }) =>
    api.get<AccessRequestsListResponse>("/access/requests", { params }),

  getRequest: (id: string) => api.get<AccessRequest>(`/access/requests/${id}`),

  createRequest: (body: { user_id: string; role_id: string; justification: string }) =>
    api.post<AccessRequest>("/access/requests", body),

  approveRequest: (id: string, comment?: string) =>
    api.post<AccessRequest>(`/access/requests/${id}/approve`, { comment }),

  rejectRequest: (id: string, comment?: string) =>
    api.post<AccessRequest>(`/access/requests/${id}/reject`, { comment }),

  withdrawRequest: (id: string) => api.delete(`/access/requests/${id}`),

  // Matrix
  getMatrix: () => api.get<PositionMatrixRow[]>("/access/matrix"),

  updateMatrix: (positionId: string, roleIds: string[]) =>
    api.patch("/access/matrix", { position_id: positionId, role_ids: roleIds }),
};

File: backend/app/models/access.py
import enum
from datetime import datetime
from uuid import UUID, uuid4

from sqlalchemy import Boolean, DateTime, Enum, ForeignKey, String, Text, func
from sqlalchemy.dialects.postgresql import UUID as PGUUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class ResourceType(str, enum.Enum):
    file_share = "file_share"
    application = "application"
    database = "database"
    api = "api"


class AccessRequestStatus(str, enum.Enum):
    pending = "pending"
    approved = "approved"
    rejected = "rejected"
    withdrawn = "withdrawn"


class Role(Base):
    __tablename__ = "roles"

    id: Mapped[UUID] = mapped_column(PGUUID(as_uuid=True), primary_key=True, default=uuid4)
    code: Mapped[str] = mapped_column(String(100), unique=True, nullable=False)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=False, default="")
    is_privileged: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    owner_user_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("users_ext.id"), nullable=True
    )

    role_permissions: Mapped[list["RolePermission"]] = relationship(
        "RolePermission", back_populates="role", cascade="all, delete-orphan"
    )
    user_roles: Mapped[list["UserRole"]] = relationship(
        "UserRole", back_populates="role", cascade="all, delete-orphan"
    )


class Permission(Base):
    __tablename__ = "permissions"

    id: Mapped[UUID] = mapped_column(PGUUID(as_uuid=True), primary_key=True, default=uuid4)
    code: Mapped[str] = mapped_column(String(100), unique=True, nullable=False)
    description: Mapped[str] = mapped_column(String(500), nullable=False, default="")

    role_permissions: Mapped[list["RolePermission"]] = relationship(
        "RolePermission", back_populates="permission", cascade="all, delete-orphan"
    )


class RolePermission(Base):
    __tablename__ = "role_permissions"

    role_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("roles.id", ondelete="CASCADE"), primary_key=True
    )
    permission_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("permissions.id", ondelete="CASCADE"), primary_key=True
    )

    role: Mapped["Role"] = relationship("Role", back_populates="role_permissions")
    permission: Mapped["Permission"] = relationship("Permission", back_populates="role_permissions")


class UserRole(Base):
    __tablename__ = "user_roles"

    id: Mapped[UUID] = mapped_column(PGUUID(as_uuid=True), primary_key=True, default=uuid4)
    user_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("users_ext.id", ondelete="CASCADE"), nullable=False
    )
    role_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("roles.id", ondelete="CASCADE"), nullable=False
    )
    granted_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    granted_by: Mapped[UUID | None] = mapped_column(PGUUID(as_uuid=True), nullable=True)
    expires_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    request_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("access_requests.id", use_alter=True, name="fk_user_role_request"),
        nullable=True,
    )

    role: Mapped["Role"] = relationship("Role", back_populates="user_roles", foreign_keys=[role_id])


class Resource(Base):
    __tablename__ = "resources"

    id: Mapped[UUID] = mapped_column(PGUUID(as_uuid=True), primary_key=True, default=uuid4)
    code: Mapped[str] = mapped_column(String(100), unique=True, nullable=False)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    type: Mapped[ResourceType] = mapped_column(
        Enum(ResourceType, name="resource_type"), nullable=False
    )
    owner_user_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("users_ext.id"), nullable=True
    )


class AccessRequest(Base):
    __tablename__ = "access_requests"

    id: Mapped[UUID] = mapped_column(PGUUID(as_uuid=True), primary_key=True, default=uuid4)
    user_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("users_ext.id"), nullable=False
    )
    role_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("roles.id"), nullable=False
    )
    justification: Mapped[str] = mapped_column(Text, nullable=False)
    status: Mapped[AccessRequestStatus] = mapped_column(
        Enum(AccessRequestStatus, name="access_request_status"),
        nullable=False,
        default=AccessRequestStatus.pending,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    decided_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    decided_by: Mapped[UUID | None] = mapped_column(PGUUID(as_uuid=True), nullable=True)
    decision_comment: Mapped[str | None] = mapped_column(Text, nullable=True)

    role: Mapped["Role"] = relationship("Role", foreign_keys=[role_id])


class PositionRoleDefault(Base):
    __tablename__ = "position_role_defaults"

    position_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("positions.id", ondelete="CASCADE"), primary_key=True
    )
    role_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("roles.id", ondelete="CASCADE"), primary_key=True
    )

    role: Mapped["Role"] = relationship("Role")


class ProcessedEvent(Base):
    __tablename__ = "processed_events"

    event_id: Mapped[UUID] = mapped_column(PGUUID(as_uuid=True), primary_key=True)
    processed_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    consumer_group: Mapped[str] = mapped_column(String(100), nullable=False)
